### PR TITLE
Common behaviour for lists with items that have address

### PR DESCRIPTION
**Pull request by karliss**
**Detailed description**

Add

* AddressableItemContextMenu - basic menu that can be used for most places that have something with address. Currently contains "show in", "copy adderess", "show X-refs", add comment. 
* AdressableItemModel - Qt Item Model with interface for asking item's address and optionally name. I considered using custom role but chose interface with virtual methods instead as it allows enforcing that interface is implemented at compile time. Manually tracking if all the models use correct role would be more fragile.
* ListDockWidget - CutterDockWidget containing tree view and optionally QuckSearch and statusbar with count. Accepts AddressableItemModel and Automatically adds AdressableItemContextMenu. This one saves a lot of duplicate code and provides more consistent behavior. A specific DockWidget implementation needs to provide model which extends AddressableItemModel if necessary customize tree view widget look and add extra context menu actions specific to this widget.

Convert following widgets to use ListDockWidget
* CommentsWidget
* ImportsWidget
* RelocsWidget
* SymbolsWidget
* SegmentsWIdget
* SectionsWIdget
* HeadersWIdget
* FunctionWidget

Some of the provided partial context menu or no context menu. Widget specific context menu entries should be still there.

Bugfixes and existing feature cleanup:
* Escape shortcut for canceling quick filter was broken due to VtablesWidget
* Selecting multiple functions in FunctionWIdget and using "Undefine" action undefines all of them
* Horizontal/Vertical mode selection now uses ActionGroup for managing exclusive selection instead of doing it manually.

**Test plan (required)**

Using context menu
* copy address
* show in
* show x-refs
  * make sure name is shown when applicable and address otherwise
* add comment
* any widget specific context menu items
* action shortcuts (with correct target)
* correct target when opening menu using keyboard

Quick filter
* Ctrl-F focuses on Quick Filter opening it if necesarry
* Quick filter works
* Escape clears the filter

Horizontal/Vertical mode for Functions and Comment widgets
* Qt displays it using Radio button style instead of checkboxes
* Tree item content switches between tree and table
* In tree mode copying adress takes function address in function widget, function or comment address in comment widget

Test at least one action in all widgets 
* CommentsWidget
* ImportsWidget
* RelocsWidget
* SymbolsWidget
* SegmentsWIdget
* SectionsWIdget
  * graphic bars show up
* HeadersWIdget
* FunctionWidget

Make sure it is possible to copy any cell in table view by clicking it and pressing Ctrl-C

** Future work  for part 2**
Deal with rest of of the widgets, probably split ListDockWidget so that TreeView with addressable item model can be used in docks that have more custom gui  like Search. 

Use of AdressableItemContextMenu in places that aren't table row, for example stack widget address and data.

Add comment should probably show existing comment for editing instead of silently replacing it. Current behavior is copied either for Comments or Functions widget.

**Closing issues**

Covers ~~most~~ ~half of #923 . ~~Need to evaluate once finished.~~

### CODE REVIEW

**Comment by ITAYC0HEN** [comment]
Sweet! I tested the PR and it looks very good, much better than before and finally - a unified solution.
I noticed just a few issues, some are probably not related to your PR but I find it worth mentioning here.

Also, I'd love to see it documented in a friendly way, in our documentation, and on API level. Basically, these additions would be used by almost any plugin developer of Cutter (that would probably want to introduce widgets based on these. You already have a nice explanation at the top of this page so it would be mostly copying and pasting I guess :))

---

1. I noticed that copying a cell with numbers in its value (and not a string or hex) doesn't work
2. Strings widget have a "Filter    ctrl+f" context menu item and other doesn't
 - do we want all the list widgets to have it?
 - anyway, clicking on this "Filter" in Strings' context menu does not put the focus to the Filter widget
3. Seems like the widgets have the [KeyboardSearch](https://doc.qt.io/qt-5/qabstractitemview.html#keyboardSearch) option enabled by default. This has a collision with one-letter shortcuts such as <kbd>s</kbd > for seeking or <kbd>x</kbd> for showing x-refs

File: src/widgets/FunctionsWidget.cpp [hunk 1]
```diff
@@ -411,64 +427,60 @@ bool FunctionSortFilterProxyModel::lessThan(const QModelIndex &left, const QMode
...
+    setObjectName("FunctionsWidget");
 
-    // Radare core found in:
-    this->main = main;
     setTooltipStylesheet();
     connect(Config(), SIGNAL(colorsUpdated()), this, SLOT(setTooltipStylesheet()));
 
 
     // leave the filter visible by default so users know it exists
     //ui->filterLineEdit->setVisible(false);
```
Comment (ITAYC0HEN) [inline src/widgets/FunctionsWidget.cpp]: Redundant comments?

**Comment by karliss** [comment]
@ITAYC0HEN 
1.  I noticed that as well. [Documentation](https://doc.qt.io/qt-5/qt.html#ItemDataRole-enum) suggests that Qt::DisplayRole should return string. Digging in code confirms that copy only  works if QString is  returned https://github.com/qt/qtbase/blob/aca43d29f8a1c90d14069ac602cf0ba7beaba300/src/widgets/itemviews/qabstractitemview.cpp#L2337 Doesn't seem like there is anything better than going through all models and wrapping numbers in `QString::number`. Seems annoying but this PR already modifies many models so I might as well do it for widgets updated in this PR and leave other widget for part 2.

2. I didn't touch string widget yet.  My personal opinion is that it's not necessary to have it context menu. I wouldn't try to look for it there. So it's main use is discoverability. But supporting Ctrl-F for all the tables and sometimes showing it by default I think is sufficient.

3. Good catch. I don't have a strong opinion for this. Shortcuts in tables and Disassembly/Graph context menu should probably match as much as possible.

File: src/widgets/FunctionsWidget.cpp (hunk 1 above)
Comment (karliss) [reply src/widgets/FunctionsWidget.cpp]: Removed in 1122f57

**Comment by ITAYC0HEN** [comment]
Before merging, what do you think should be done with point 3? Should we disable KeyboardSearch?

**Comment by karliss** [comment]
Show in works as designed but due to how global seek synchronization works it isn't perfect.

If target widget isn't synchronized only target seek will be changed. I changed the default synchronization state to disabled for new widgets as to avoid unexpected seek change in current widget. Didn't touch show in existing widget as that can cause unexpected widget desynchronization which can cause different kind of problems.

**Pull request merged**